File: framework/navigation-dxa-module/src/main/java/com/sdl/webapp/navigation/model/NavigationSection.java
package com.sdl.webapp.navigation.model;

import com.sdl.webapp.common.api.mapping.semantic.annotations.SemanticEntity;
import com.sdl.webapp.common.api.mapping.semantic.annotations.SemanticProperty;
import com.sdl.webapp.common.api.model.entity.AbstractEntityModel;

import java.util.ArrayList;
import java.util.List;

import static com.sdl.webapp.common.api.mapping.semantic.config.SemanticVocabulary.SDL_CORE;

/**
 * Navigation Section
 *
 * @author nic
 */
@SemanticEntity(entityName = "NavigationSection", vocabulary = SDL_CORE, prefix = "e", public_ = false)
public class NavigationSection extends AbstractEntityModel {

    @SemanticProperty("e:mainItem")
    private NavigationItem mainItem;

    @SemanticProperty("e:items")
    private List<NavigationItem> items;

    // TODO: Have this configurable in metadata
    private int cols = 4;

    public NavigationItem getMainItem() {
        return mainItem;
    }

    public List<NavigationItem> getItems() {
        return items;
    }

    public List<List<NavigationItem>> getItemColumns() {
        List<List<NavigationItem>> columns = new ArrayList<>();
        if ( items != null ) {
            List<NavigationItem> column = new ArrayList<>();
            columns.add(column);
            int i = 0;
            for (NavigationItem item : items) {
                column.add(item);
                i++;
                if (i % cols == 0) {
                    column = new ArrayList<>();
                    columns.add(column);
                }
            }
        }
        return columns;
    }
}

File: dxa.net/ecommerce-framework-dotnet/SDL.DXA.Modules.Navigation/Models/NavigationSection.cs
using Sdl.Web.Common.Models;

using System.Collections.Generic;

namespace SDL.DXA.Modules.Navigation.Models
{
    /// <summary>
    /// Navigation Section
    /// </summary>
    [SemanticEntity(Vocab = CoreVocabulary, EntityName = "NavigationSection", Prefix = "e")]
    public class NavigationSection : EntityModel
    {
        // TODO: Have this configurable in metadata
        const int COLS = 4;

        [SemanticProperty("e:mainItem")]
        public NavigationItem MainItem { get; set; }

        [SemanticProperty("e:items")]
        public IList<NavigationItem> Items { get; set; }

        /// <summary>
        /// Get items in columns
        /// </summary>
        /// <returns></returns>
        public IList<IList<NavigationItem>> GetItemColumns()
        {
            var columns = new List<IList<NavigationItem>>();
            if (Items != null)
            {
                var column = new List<NavigationItem>();
                columns.Add(column);
                int i = 0;
                foreach (var item in Items)
                {
                    column.Add(item);
                    i++;
                    if (i % COLS == 0)
                    {
                        column = new List<NavigationItem>();
                        columns.Add(column);
                    }
                }
            }
            return columns;
        }
    }
}